kythe/rust/fuchsia_extractor/README.md

```diff
@@ -1,9 +1,168 @@
 
 [fx]: https://www.fuchsia.dev
+[gn]: https://gn.googlesource.com/gn/
 
-Fuchsia build system is a bit specific, and is custom built on top of [GN][gn].
+Technically, this extractor produces `kzip` files as any other extractor does,
+and the `kzip` file specification is well known.  However, in practice, the
+kzip specification is open for interpretation (for example, what if you need
+to inject a specific extra file to a compilation unit?).  That in turn can
+lead to mismatches in the interpretations that would break the ingestion
+pipeline.
+
+This document is an attempt to lay out any such assumptions for
+`fuchsia_extractor` going forward.
+
+# Specifics of Fuchsia compilation
+
+Fuchsia is compiled with [GN][gn].  However, due to the size of the code base,
+the GN generator scripts are heavily customized and templatized in a specific
+way, so it could be said that Fuchsia uses a custom purpose-built build system.
+
+Fuchsia source code is spread across many repositories, some of which can be
+seen at https://fuchsia.googlesource.com.  When the source code is checked out,
+a custom utility `fx jiri` is used (which is to say, a custom wrapper over the
+`jiri` program) to check out all the repositories into its appropriate place
+underneath a common source root directory, which we will call `$FUCHSIA_DIR`
+after the environment variable that is usually used to refer to it in scripts.
+
+Fuchsia code is compiled in a separate "output" directory.  However, this
+directory is created underneath `$FUCHSIA_DIR`, and each build configuration,
+for a "product-dot-board" combination such as `terminal.x64`, gets a separate
+build output directory.  The build configuration `terminal.x64` gets an output
+directory in `$FUCHSIA_DIR/out/terminal.x64`, though this last path component
+can be given a different name by the developer.  Let us call this directory
+`$FUCHSIA_OUT` for short.
+
+Multiple code corpora exist in `$FUCHSIA_DIR` and `$FUCHSIA_OUT`.  There is C
+and C++ code, Rust code, go code, test fixtures, dart and flutter code, and
+dedicated IDL generated code for languages called FIDL and Banjo.
+
+When filling out compilation unit data, we need to ensure that when we refer to
+files, their names are expressed relative to some of those directories, and
+then to correct ones.  Furthermore, sometimes relative paths are needed with
+respect to the corpus root, and sometimes with respect to the build directory.
+For example, generated code lives in `$FUCHSIA_OUT/gen`.  Object code, such as
+`.o` files, live in `$FUCHSIA_OUT/obj`, and there are about a dozen of such
+different corpus roots.
+
+Even further, though, even as the relationships between these directories are fixed,
+infrastructure builds may take liberties with the arrangement of these directories.
+This means that we need to keep options open to retarget these directories
+as the developer sees fit for their use.  Hence the many flags.
```

Comment:
This is generally accomplished via `vnames.json` files.

Reply:
Ack.